vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +165,108 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
 	}
 	return delay, nil
 }
+
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight uint64,
+	parentBlockTime,
+	blockTime time.Time,
+) (ids.NodeID, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		slot             = uint64(blockTime.Sub(parentBlockTime) / WindowDuration)
+		seed             = blockHeight ^ bits.Reverse64(slot) ^ w.chainSource
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+	)
+
+	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	source.Seed(seed)
+	indices, err := sampler.Sample(1)
+	if err != nil {
+		return ids.EmptyNodeID, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+	}
+	return validators[indices[0]].id, nil
+}
+
+func (w *windower) MinDelayForProposer(
+	ctx context.Context,
+	chainHeight,
+	pChainHeight uint64,
+	parentBlockTime time.Time,
+	nodeID ids.NodeID,
+	startTime time.Time,
+) (time.Duration, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return 0, err
+	}
+
+	var (
+		slot = uint64(startTime.Sub(parentBlockTime) / WindowDuration)
+
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+		sampler          = sampler.NewDeterministicWeightedWithoutReplacement(source)
+	)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return 0, err
+	}
+
+	for time.Duration(slot)*WindowDuration < MaxLookAheadWindow {
+		seed := chainHeight ^ bits.Reverse64(slot) ^ w.chainSource
+		source.Seed(seed)
```

Comment:
To ask a dumb question, re-calling `Seed` here is equivalent to a fresh initialization right?

Reply:
in principle we should re create the sampler with the new seed. However Stephen mentioned that this is fine for the way the sampler is implemented, and it is more efficient